numpy/core/function_base.py

```diff
@@ -102,9 +102,15 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
         if step == 0:
             # Special handling for denormal numbers, gh-5437
             y /= div
-            y *= delta
-        else:
+            step = delta
+
+        if getattr(step, '__array_priority__', 0) <= 0:
+            # Inplace is faster, but we should only do it if the step is not
+            # something that wants to override array, since in-place is broken
+            # for that case; this allows subclasses to use linspace; gh-7142
             y *= step
+        else:
+            y = y*step
     else:
         # 0 and 1 item long sequences have an undefined step
```

Comment:
This was how the code worked pre 1.10: with a regular multiplication, `ndarray` will defer to `step.__rmul__` if needed, thus allowing the subclass to determine the class of the result (for the particular problem here, `step` is an astropy `Quantity`).